<!DOCTYPE HTML>
<html lang="en" class="light sidebar-visible" dir="ltr">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Bioinformatics with Rust</title>
        <meta name="robots" content="noindex">


        <!-- Custom HTML head -->

        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff">

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" id="highlight-css" href="highlight.css">
        <link rel="stylesheet" id="tomorrow-night-css" href="tomorrow-night.css">
        <link rel="stylesheet" id="ayu-highlight-css" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->


        <!-- Provide site root and default themes to javascript -->
        <script>
            const path_to_root = "";
            const default_light_theme = "light";
            const default_dark_theme = "navy";
        </script>
        <!-- Start loading toc.js asap -->
        <script src="toc.js"></script>
    </head>
    <body>
    <div id="mdbook-help-container">
        <div id="mdbook-help-popup">
            <h2 class="mdbook-help-title">Keyboard shortcuts</h2>
            <div>
                <p>Press <kbd>←</kbd> or <kbd>→</kbd> to navigate between chapters</p>
                <p>Press <kbd>S</kbd> or <kbd>/</kbd> to search in the book</p>
                <p>Press <kbd>?</kbd> to show this help</p>
                <p>Press <kbd>Esc</kbd> to hide this help</p>
            </div>
        </div>
    </div>
    <div id="body-container">
        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                let theme = localStorage.getItem('mdbook-theme');
                let sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            const default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? default_dark_theme : default_light_theme;
            let theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            const html = document.documentElement;
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add("js");
        </script>

        <input type="checkbox" id="sidebar-toggle-anchor" class="hidden">

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            let sidebar = null;
            const sidebar_toggle = document.getElementById("sidebar-toggle-anchor");
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            sidebar_toggle.checked = sidebar === 'visible';
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <!-- populated by js -->
            <mdbook-sidebar-scrollbox class="sidebar-scrollbox"></mdbook-sidebar-scrollbox>
            <noscript>
                <iframe class="sidebar-iframe-outer" src="toc.html"></iframe>
            </noscript>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle">
                <div class="sidebar-resize-indicator"></div>
            </div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <label id="sidebar-toggle" class="icon-button" for="sidebar-toggle-anchor" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </label>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="default_theme">Auto</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search (`/`)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="/ s" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">Bioinformatics with Rust</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="introduction"><a class="header" href="#introduction">Introduction</a></h1>
<p>Welcome to Bioinformatics with Rust! An unofficial book aimed at introducing the Rust programming language for bioinformatic applications. Before we start:</p>
<ul>
<li>
<p>This book is <strong>not</strong> in any way, shape or form, an official introduction to neither the Rust programming language, nor bioinformatics. For an introduction to Rust, please visit <a href="prefix/../../suffix/1_resources.html#resources">resources</a>.</p>
</li>
<li>
<p>The <strong>purpose</strong> of this book is to simply give some inspiration on how the Rust programming language
can be used to create fast and memory safe bioinformatics pipelines.</p>
</li>
<li>
<p>For any issues related to this book, please file a <a href="https://github.com/OscarAspelin95/bioinformatics_with_rust/issues">GitHub issue</a>.</p>
</li>
<li>
<p>Contributions are very welcome!</p>
</li>
</ul>
<div style="break-before: page; page-break-before: always;"></div><h1 id="prerequisites"><a class="header" href="#prerequisites">Prerequisites</a></h1>
<p>If you have never heard of the Rust programming language or bioinformatics, this book is probably not for you.</p>
<p>If you have a lot of experience with both Rust and bioinformatics, this book is probably not for you.</p>
<p>However, if you know a bit about bioinformatics and have worked with programming languages such as Python or C++ then this book will give you an introduction to the Rust ecosystem for building bioinformatic pipelines!</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="getting-started"><a class="header" href="#getting-started">Getting Started</a></h1>
<p>First and foremost, we need to install Rust and its package manager Cargo. The easiest way to do this is to use <a href="https://rustup.rs/">Rustup</a> and follow the installation instructions.</p>
<p>Second, we need a code editor. For running Rust in hardcode mode, use notepad. However, then we are missing some important stuff, like syntax highlighting, code formatting and much more. Here are some examples of code editors that can get the job done:</p>
<ul>
<li><a href="https://code.visualstudio.com/">VScode</a> - Easy to use with lots of plugins to make your life easier.</li>
<li><a href="https://helix-editor.com/">Helix</a> - A text editor written in Rust!</li>
<li><a href="https://www.vim.org/">Vim</a> - For hardcore programmers.</li>
<li><a href="https://neovim.io/">NeoVim</a> - For modern, hardcore programmers.</li>
</ul>
<p>If using VScode, there are some handy plugins that will make your Rust programming journey more pleasant:</p>
<ul>
<li><a href="https://github.com/rust-lang/rust-analyzer">Rust-analyzer</a> - Trust me, you want this.</li>
<li><a href="https://github.com/astral-sh/ruff">Ruff</a> - Code formatter and linter written in Rust.</li>
<li><a href="https://marketplace.visualstudio.com/items?itemName=tamasfe.even-better-toml">Even Better TOML</a> - TOML file support.</li>
<li><a href="https://github.com/features/copilot">GitHub Copilot</a> - Copilot is relatively good at Rust, but use with caution and be aware of hallucinations.</li>
<li><a href="https://marketplace.visualstudio.com/items?itemName=mechatroner.rainbow-csv">Rainbow CSV</a> - If working with CSV files, this plugin provides highlighting and interactive querying.</li>
</ul>
<div style="break-before: page; page-break-before: always;"></div><h1 id="example-code"><a class="header" href="#example-code">Example Code</a></h1>
<div style="break-before: page; page-break-before: always;"></div><h1 id="rust-code"><a class="header" href="#rust-code">Rust Code</a></h1>
<pre><pre class="playground"><code class="language-rust edition2024">/// This is just some example code.
fn main(){
    let number: usize = 5;
    println!("{number}");
}</code></pre></pre>
<pre><pre class="playground"><code class="language-rust edition2024"><span class="boring">// This is a comment that is shown when clicking the eye icon.
</span>/// This is some other example code.
fn main(){
    let number: usize = 5;
    println!("{number}");
}</code></pre></pre>
<p>See <a href="https://rust-lang.github.io/mdBook/format/mdbook.html">this link</a> for including a separate .rs file in the book.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="nucleotides"><a class="header" href="#nucleotides">Nucleotides</a></h1>
<p>When dealing with DNA sequences, nucleotides are everything. Fundamentally, we are usually dealing with four canonical bases:</p>
<ul>
<li>A - Adenosine.</li>
<li>C - Cytosine.</li>
<li>G - Guanine.</li>
<li>T - Thymine.</li>
</ul>
<p>However, there are some important concepts to be aware of:</p>
<ul>
<li>
<p><strong>Soft masking</strong>: A lower case nucleotide indicates <em>soft masking</em> and is used to indicate a soft clipped alignment or a region which is low complexity or repetitive.</p>
<p>Example of soft masked, highly repetitive sequence:</p>
<ul>
<li>ATCGatatatatatatatatatat[...]AGCGAGT</li>
</ul>
<p>Example of soft clipped alignment:</p>
  <pre>
  * AAAGTGCCAGTGACGCTTagtcgatcgatg
    ||||||||||||||||||
    AAAGTGCCAGTGACGCTT
  <pre>

</li>
<li>
<p><strong>Hard masking</strong>: A capital "N" indicates <em>hard masking</em>. This means there is probably a base here, but we don't know exactly what it is. This is usually for indicating uncertainty or gaps in a sequence.</p>
</li>
</ul>
<div style="break-before: page; page-break-before: always;"></div><h1 id="create-nucleotide-sequence"><a class="header" href="#create-nucleotide-sequence">Create nucleotide sequence</a></h1>
<p>To start off, we just create a nucleotide sequence as a String.</p>
<pre><pre class="playground"><code class="language-rust edition2024">fn main(){
    let nt_string: String = "ACGT".to_string();

    println!("{nt_string}");
}
</code></pre></pre>
<p>However, usually when reading nucleotide sequences from a FASTA/Q parser, we get it in binary form &amp;[u8] which is a more convenient format.</p>
<pre><pre class="playground"><code class="language-rust edition2024">fn main(){
    let nt_string_binary: &amp;[u8] = b"ACGT";

    // Here, we need to print in debug mode.
    println!("{:?}", nt_string_binary);
}</code></pre></pre>
<p>Run the code and examine the output. We get a bunch of numbers. This is the ASCII representation of our nucleotides, where A/T/C/G corresponds to a 8-bit representation. For more information, visit <a href="https://www.ascii-code.com/">this link</a>.</p>
<p>We can check that the following representations are equivalent:</p>
<pre><pre class="playground"><code class="language-rust edition2024">fn main(){
    // Note that here we use single quotes
    // because we are dealing with single chars.
    assert_eq!(b'A', 65);
    assert_eq!(b'C', 67);
    assert_eq!(b'G', 71);
    assert_eq!(b'T', 84);
}</code></pre></pre>
<p>Onwards, we'll mostly deal with nucleotides in binary form.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="count-nucleotides"><a class="header" href="#count-nucleotides">Count nucleotides</a></h1>
<p>We can implement a simple script for calculating nucleotides.</p>
<pre><pre class="playground"><code class="language-rust edition2024">fn main(){
    let nt_string: &amp;[u8] = b"AAACGTGACGT";

    // Variables must be mutable for us to increment the count.
    let mut a = 0;
    let mut c = 0;
    let mut g = 0;
    let mut t = 0;

    // Iterate over each nucleotide.
    nt_string.iter().for_each(|nt| match nt {
        b'A' =&gt; a += 1,
        b'C' =&gt; c += 1,
        b'G' =&gt; g += 1,
        b'T' =&gt; t += 1,
        // We only allow canonical, non masked bases.
        _ =&gt; panic!("Invalid nt {nt}"),
    });

    assert_eq!(a + c + g + c, nt_string.len());
    println!("A: {a}, C: {c}, G: {g}, T: {t}");
}</code></pre></pre>
<p>This approach is not that elegant, because we initialize four different variables, one for each nucleotide. An alternative approach is to use a HashMap.</p>
<pre><pre class="playground"><code class="language-rust edition2024">use std::collections::HashMap;

fn main() {
    let nt_string: &amp;[u8] = b"AAACGTGACGT";

    // HashMap must be mutable for us to modify it.
    let mut map: HashMap&lt;&amp;u8, usize&gt; = HashMap::new();

    // Iterate over each nucleotide.
    nt_string.iter().for_each(|nt| match nt {
        // If we have a canonical nucleotide, we bind it to the variable c.
        c @ (b'A' | b'C' | b'G' | b'T') =&gt; match map.contains_key(c) {
            // If nucleotide is already in HashMap, increment its count.
            true =&gt; {
                let v = map.get_mut(c).unwrap();
                *v += 1;
            }
            // Otherwise, add it.
            false =&gt; {
                map.insert(c, 1);
            }
        },
        _ =&gt; panic!("Invalid nt {nt}"),
    });

    assert_eq!(map.values().sum::&lt;usize&gt;(), nt_string.len());
    println!("{:?}", map);
}</code></pre></pre>
<p>Run the code and inspect the output. The resulting HashMap will have the ASCII encoded nucleotides as keys.</p>
<p>Note that there are lots of alternative solutions and further optimizations we can do.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="reverse-complement"><a class="header" href="#reverse-complement">Reverse complement</a></h1>
<p>Next, we cover a fundamental topic, which is reverse complementing. Why is this important?</p>
<p>DNA is (generally) double stranded, where bases are paired:</p>
<ul>
<li>A pairs with T and vice versa.</li>
<li>G pairs with C and vice versa.</li>
</ul>
<pre>
5' [...]ACGAGCTTTGTGACGCGATGCGACGAGCTGCAGCGT[...] 3'
3' [...]TGCTCGAAACACTGCGCTACGCTGCTCGACGTCGCA[...] 5'
</pre>
<p>Pretend this is a bacterial genome we want to sequence. Before sequencing, we need to separate the strands and break this molecule into smaller pieces. When doing this, we don't know which pieces are from which strand.</p>
<p>When we want to align the pieces back to a reference sequence (which is defined in the 5' to 3' direction), we need to take both strands into consideration. Otherwise, we loose out on information. We do this by reverse complementing, in which we first reverse the sequence, and then replace each base with the corresponding matching base.</p>
<p>For example, the reverse complement of ATCG would be CGAT.</p>
<p>We can relatively easily write a Rust function for reverse complementing a nucleotide sequence:</p>
<pre><pre class="playground"><code class="language-rust edition2024">fn reverse(nt: char) -&gt; char {
    match nt {
        'A' =&gt; 'T',
        'C' =&gt; 'G',
        'G' =&gt; 'C',
        'T' =&gt; 'A',
        _ =&gt; panic!("Invalid nt {nt}"),
    }
}

fn reverse_complement(nt_string: &amp;'static str) -&gt; String {
    let rev_comp: String = nt_string
        // Iterate over each character.
        .chars()
        // Reverse the iteration order.
        .rev()
        .map(|nt| {
            return reverse(nt);
        })
        .collect();

    return rev_comp;
}
fn main() {
    // Note that we use strings and not &amp;[u8].
    // &amp;[u8] will be covered in a future topic.
    assert_eq!(reverse_complement("AAA"), "TTT");
    assert_eq!(reverse_complement("GGG"), "CCC");
    assert_eq!(reverse_complement("ATCG"), "CGAT");
    assert_eq!(reverse_complement("ACACGT"), "ACGTGT");
}</code></pre></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="encoding"><a class="header" href="#encoding">Encoding</a></h1>
<p>Being able to encode and decode nucletides is a vital part of writing high performance bioinformatic code. What it means is essentially converting nucleotides into a more compact form. There are multiple ways of doing nucleotide encoding. However if we assume we only have to deal with A/C/G/T, then there is a straightforward way for this:</p>
<ul>
<li>A is encoded as 0 (binary 00).</li>
<li>C is encoded as 1 (binary 01).</li>
<li>G is encoded as 2 (binary 10).</li>
<li>T is encoded as 3 (binary 11).</li>
</ul>
<p>The advantages of this approach are:</p>
<ul>
<li>Each nucleotides only takes up 2 bits.</li>
<li>Reverse complementing a base is as easy as:
<ul>
<li>rev_nt = 3 - nt</li>
</ul>
</li>
<li>With some bit-shifting, we can very efficiently generate kmers from our sequences (covered in a later topic).</li>
</ul>
<p>The following code is just a very straightforward encoding/decoding protocol. However, this enables us to do some more advanced stuff in future topics.</p>
<pre><pre class="playground"><code class="language-rust edition2024">/// Convert ASCII to our own 2-bit encoded nt.
fn encode(nt_decoded: u8) -&gt; u8 {
    match nt_decoded {
        b'A' =&gt; 0,
        b'C' =&gt; 1,
        b'G' =&gt; 2,
        b'T' =&gt; 3,
        _ =&gt; panic!("Invalid nt {nt_decoded}"),
    }
}

/// Convert our own 2-bit encoded nt to ASCII.
fn decode(nt_encoded: u8) -&gt; u8 {
    match nt_encoded {
        0 =&gt; b'A',
        1 =&gt; b'C',
        2 =&gt; b'G',
        3 =&gt; b'T',
        _ =&gt; panic!("Invalid nt {nt_encoded}"),
    }
}

/// Reverse complement an ASCII base.
fn reverse(nt: u8) -&gt; u8 {
    return decode(3 - encode(nt));
}

/// Reverse complement a nucleotide sequence.
fn reverse_complement(nt_string: &amp;[u8]) -&gt; Vec&lt;u8&gt; {
    nt_string.iter().rev().map(|nt| reverse(*nt)).collect()
}

fn main() {
    // Reverse complement a single nucleotide.
    assert_eq!(reverse(b'A'), b'T');
    assert_eq!(reverse(b'T'), b'A');
    assert_eq!(reverse(b'C'), b'G');
    assert_eq!(reverse(b'G'), b'C');

    // We can also reverse complement a nucleotide sequence.
    assert_eq!(reverse_complement(b"AAAA").as_slice(), b"TTTT");
    assert_eq!(reverse_complement(b"ATCG").as_slice(), b"CGAT");
}
</code></pre></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="basics-of-alignment"><a class="header" href="#basics-of-alignment">Basics Of Alignment</a></h1>
<h2 id="introduction-1"><a class="header" href="#introduction-1">Introduction</a></h2>
<p>In bioinformatics, alignment is the process of determining how well biological sequences match to each other. Usually, we refer to the sequences as the <em>query</em> and <em>subject</em> respectively. For simplicity, we'll assume that the query and subject both are single sequences.</p>
<p>There are three important alignment features to understand:</p>
<ul>
<li>Match.</li>
<li>Mismatch.</li>
<li>Insertion/Deletion.</li>
</ul>
<p>In the following alignment, matches are shown with a vertical bar (|), mismatches as asterisks (*) and insertions or deletions as hyphens (-).</p>
<pre>
query   AGCGACTCGTGCTCGA-CTT
        | |||||||*|||||| |||
subject A-CGACTCGAGCTCGAGCTT
</pre>
<h2 id="definitions"><a class="header" href="#definitions">Definitions</a></h2>
<ul>
<li>
<p><em>Query length</em> = The length of the query sequence.</p>
</li>
<li>
<p><em>Subject length</em> = The length of the subject sequence.</p>
</li>
<li>
<p><em>Alignment length</em> = The length of the aligned part between the query and subject.</p>
</li>
<li>
<p><em>Percent Identity</em> = 100 * (num_matches / alignment_length).</p>
</li>
<li>
<p><em>Fraction Aligned (Query)</em> = query_length / alignment_length (how much of the query is aligned).</p>
</li>
<li>
<p><em>Fraction Aligned (Subject)</em> = subject_length / alignment_length (how much of the subject is aligned).</p>
</li>
</ul>
<h2 id="types-of-alignments"><a class="header" href="#types-of-alignments">Types Of Alignments</a></h2>
<p>There are three basic types of alignments:</p>
<ul>
<li>
<p><em>Global Alignment</em> - Aligns the entire query against the entire subject. Suitable if query and subject are of similar length, or one expects the entire query to align against the entire subject. An example is aligning two very similar genomes of roughly the same length.</p>
</li>
<li>
<p><em>Semi Global Alignment</em> - Fully aligns the shorter of query/subject. An example is trying to align a gene (shorter) against an entire genome (longer).</p>
</li>
<li>
<p><em>Local Alignment</em> - Allows partial alignment of the query against the subject.</p>
</li>
</ul>
<div style="break-before: page; page-break-before: always;"></div><h1 id="hamming-distance"><a class="header" href="#hamming-distance">Hamming Distance</a></h1>
<p>The <a href="https://en.wikipedia.org/wiki/Hamming_distance">Hamming distance</a> is defined as the number of positions between two strings of equal length that are different. Hence, it measures the number of substitutions needed to convert one string to the other.</p>
<p>In the example below, the Hamming distance is 1.</p>
<pre>
query   ATCTACCG
        |||||*||
subject ATCGATCG
</pre>
<p>We can pretty easily write a Rust function for calculating the Hamming distance:</p>
<pre><pre class="playground"><code class="language-rust edition2024">use std::iter::zip;

fn hamming_distance(query: &amp;str, subject: &amp;str) -&gt; usize {
    assert_eq!(query.len(), subject.len());

    let mut distance = 0;

    for (query_nt, subject_nt) in zip(query.chars(), subject.chars()) {
        if query_nt != subject_nt {
            distance += 1;
        }
    }

    return distance;
}
fn main() {
    assert_eq!(hamming_distance("ATCG", "ATCG"), 0);
    assert_eq!(hamming_distance("ATCG", "TTCG"), 1);
    // Our function can actually handle non-nucleotide strings.
    assert_eq!(hamming_distance("Hello", "Heiol"), 3);
}</code></pre></pre>
<p>By implementing the Hamming distance, we have kind of almost implemented a <em>global aligner</em> (that only handles matches and mismatches).</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="resources"><a class="header" href="#resources">Resources</a></h1>
<p>The following sections list some really good resources. Feel free to browse and check them out!</p>
<p>Here are some resources related to the Rust programming language:</p>
<ul>
<li><a href="https://doc.rust-lang.org/book/">The Rust Programming Language</a> - Official Rust book.</li>
<li><a href="https://practice.course.rs/">Rust By Practice</a> - Practice Rust.</li>
<li><a href="https://doc.rust-lang.org/nomicon/intro.html">The Rustonomicon</a> - Even more Rust!</li>
</ul>
<div style="break-before: page; page-break-before: always;"></div><h1 id="awesome-rust-crates"><a class="header" href="#awesome-rust-crates">Awesome Rust crates</a></h1>
<p>General purpose awesome Rust crates!
For a mega list of even more crates, see <a href="https://github.com/rust-unofficial/awesome-rust">Awesome Rust</a>.</p>
<ul>
<li><a href="https://docs.rs/clap/latest/clap/">Clap</a> - Command line argument parsing.</li>
<li><a href="https://docs.rs/rstest/latest/rstest/">Rstest</a> - Rust test fixtures.</li>
<li><a href="https://docs.rs/rayon/latest/rayon/">Rayon</a> - Multithreading library for iterators.</li>
<li><a href="https://docs.rs/serde/latest/serde/">Serde</a> - Serialization/Deserialization.</li>
<li><a href="https://docs.rs/thiserror/latest/thiserror/">Thiserror</a> - Easily create custom error types.</li>
<li><a href="https://docs.rs/log/latest/log/">Log</a> and <a href="https://docs.rs/simple_logger/latest/simple_logger/">SimpleLogger</a> - Switch those pesky println! macros for proper logging.</li>
<li><a href="https://docs.rs/flate2/latest/flate2/">Flate2</a> - Compression/Decompression library.</li>
<li><a href="https://docs.rs/bindgen/latest/bindgen/">Bindgen</a> - Rust bindings for C and C++.</li>
<li><a href="https://docs.rs/polars/latest/polars/">Polars</a> - Blazingly fast dataframes. <strong>NOTE</strong> - Using polars with native Rust can be a bit cumbersome. An alternative is to use the <a href="https://docs.pola.rs/api/python/stable/reference/index.html">Python bindings</a>.</li>
<li><a href="https://docs.rs/pyo3/latest/pyo3/">Pyo3</a> - Generate Rust bindings to Python or vice versa.</li>
<li><a href="https://github.com/plotly/plotly.rs">Plotly-rs</a> - Rust bindings for the popular Plotly plotting library.</li>
<li><a href="https://docs.rs/linfa/latest/linfa/">Linfa</a> - Closely resembles Pythons scikit-learn for Machine Learning applications.</li>
<li><a href="https://github.com/DioxusLabs/dioxus">Dioxus</a> - Fullstack framework in Rust that resembles React. Build you own (bioinformatic) web or desktop applications!</li>
</ul>
<div style="break-before: page; page-break-before: always;"></div><h1 id="awesome-bioinformatic-tools"><a class="header" href="#awesome-bioinformatic-tools">Awesome bioinformatic tools</a></h1>
<p>Alignment related:</p>
<ul>
<li><a href="https://github.com/lh3/minimap2">Minimap2</a> - Pairwise aligner. Written by the legendary <a href="https://github.com/lh3">Heng Li</a>. The go-to for Oxford Nanopore and PacBio data.</li>
<li><a href="https://github.com/lh3/bwa">BWA-MEM</a> - Pairwise aligner suitable for Illumina data.</li>
<li><a href="https://github.com/jeffdaily/parasail">Parasail</a> - General purpose pairwise aligner.</li>
<li><a href="https://docs.rs/parasail-rs/latest/parasail_rs/">Parasail-rs</a> - Rust bindings for the parasail library.</li>
<li><a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> - The usual go-to for local sequence alignment.</li>
</ul>
<p>Assembly related:</p>
<ul>
<li><a href="https://github.com/mikolmogorov/Flye">Flye</a> - Genome assembler for Oxford Nanopore or PacBio data.</li>
<li><a href="https://github.com/loneknightpy/idba">IDBA</a> - Illumina specific genome assembler.</li>
<li><a href="https://github.com/ablab/spades">SPAdes</a> - Genome assembler suitable for Illumina or IonTorrent data.</li>
</ul>
<p>Variant calling related:</p>
<ul>
<li><a href="https://github.com/HKU-BAL/Clair3">Clair3</a> - Variant caller suitable for Illumina, Oxford Nanopore or PacBio data.</li>
<li><a href="https://github.com/nanoporetech/medaka">Medaka</a> - Variant caller and polishing tool specifically for Oxford Nanopore data.</li>
<li><a href="https://github.com/freebayes/freebayes">Freebayes</a> - Variant caller suitable for Illumina and IonTorrent data. Questionable use for Oxford Nanopore data.</li>
</ul>
<p>Misc:</p>
<ul>
<li><a href="https://github.com/samtools/samtools">SAMtools</a> - SAM file manipulation.</li>
<li><a href="https://github.com/samtools/bcftools">BCFtools</a> - VCF file manipulation.</li>
<li><a href="https://github.com/lh3/kmer-cnt">Kmer-cnt</a> - Several kmer counting algorithms.</li>
</ul>
<p>Tools written in Rust:</p>
<ul>
<li><a href="https://docs.rs/needletail/latest/needletail/">Needletail</a> - Parsing and processing FASTA/Q files.</li>
<li><a href="http://docs.rs/bio/latest/bio/">Bio</a> - General purpose bioinformatic tool for alignment, file processing and much more.</li>
<li><a href="https://github.com/bluenote-1577/sylph">Sylph</a> - Metagenomic classification tool.</li>
<li><a href="https://docs.rs/rust-htslib/latest/rust_htslib/">Rust Htslib</a> - Rust bindings for Htslib.</li>
<li><a href="https://github.com/nextstrain/nextclade">NextClade</a> - Virus specific tool for alignment, SNP calling, clade assignment and more.</li>
</ul>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->


                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">

            </nav>

        </div>

        <!-- Livereload script (if served using the cli tool) -->
        <script>
            const wsProtocol = location.protocol === 'https:' ? 'wss:' : 'ws:';
            const wsAddress = wsProtocol + "//" + location.host + "/" + "__livereload";
            const socket = new WebSocket(wsAddress);
            socket.onmessage = function (event) {
                if (event.data === "reload") {
                    socket.close();
                    location.reload();
                }
            };

            window.onbeforeunload = function() {
                socket.close();
            }
        </script>



        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->

        <script>
        window.addEventListener('load', function() {
            window.setTimeout(window.print, 100);
        });
        </script>

    </div>
    </body>
</html>
